src/data/setoid.lean

```diff
@@ -0,0 +1,488 @@
+lemma eq_iff_classes_eq' {r₁ r₂ : setoid α} :
+  r₁ = r₂ ↔ ∀ x, {y | r₁.rel x y} = {y | r₂.rel x y} :=
+⟨λ h x, h ▸ rfl, λ h, ext' $ λ x, (set.ext_iff _ _).1 $ h x⟩
+
+/-- Two equivalence relations are equal iff their equivalence classes are equal. -/
+lemma classes_inj {r₁ r₂ : setoid α} :
+  r₁ = r₂ ↔ r₁.classes = r₂.classes :=
+⟨λ h, h ▸ rfl, λ h, ext' $ λ a b,
+  ⟨λ h1, let ⟨w, hw⟩ := show _ ∈ r₂.classes, by rw ←h; exact r₁.mem_classes a in
+      r₂.trans' (show a ∈ {x | r₂.rel x w}, from hw ▸ r₁.refl' a) $
+        r₂.symm' (show b ∈ {x | r₂.rel x w}, by rw ←hw; exact r₁.symm' h1),
+   λ h1, let ⟨w, hw⟩ := show _ ∈ r₁.classes, by rw h; exact r₂.mem_classes a in
+      r₁.trans' (show a ∈ {x | r₁.rel x w}, from hw ▸ r₂.refl' a) $
+        r₁.symm' (show b ∈ {x | r₁.rel x w}, by rw ←hw; exact r₂.symm' h1)⟩⟩
+
+/-- The empty set is not an equivalence class. -/
+lemma empty_not_mem_classes {r : setoid α} : ∅ ∉ r.classes :=
+λ ⟨y, hy⟩, set.not_mem_empty y $ hy.symm ▸ r.refl' y
+
+/-- Equivalence classes partition the type. -/
+lemma classes_eqv_classes {r : setoid α} :
+  ∀ a, ∃ b ∈ r.classes, a ∈ b ∧ ∀ b' ∈ r.classes, a ∈ b' → b = b' :=
+λ a, ⟨{x | r.rel x a}, r.mem_classes a,
+  ⟨r.refl' a, λ s ⟨y, h⟩ ha, by rw h at *; ext;
+    exact ⟨λ hx, r.trans' hx ha, λ hx, r.trans' hx $ r.symm' ha⟩⟩⟩
+
+/-- If x ∈ α is in 2 equivalence classes, the equivalence classes are equal. -/
+lemma eq_of_mem_classes {r : setoid α} {x b} (hc : b ∈ r.classes)
+  (hb : x ∈ b) {b'} (hc' : b' ∈ r.classes) (hb' : x ∈ b') : b = b' :=
+eq_of_mem_eqv_class classes_eqv_classes hc hb hc' hb'
+
+/-- The elements of a set of sets partitioning α are the equivalence classes of the
+    equivalence relation defined by the set of sets. -/
+lemma eq_eqv_class_of_mem {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b')
+  {s y} (hs : s ∈ c) (hy : y ∈ s) : s = {x | rel (mk_classes c H) x y} :=
+set.ext $ λ x,
+  ⟨λ hs', symm' (mk_classes c H) $ λ b' hb' h', eq_of_mem_eqv_class H hs hy hb' h' ▸ hs',
+   λ hx, let ⟨b', hc', hb', h'⟩ := H x in
+     (eq_of_mem_eqv_class H hs hy hc' $ hx b' hc' hb').symm ▸ hb'⟩
+
+/-- The equivalence classes of the equivalence relation defined by a set of sets
+    partitioning α are elements of the set of sets. -/
+lemma eqv_class_mem {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b') {y} :
+  {x | rel (mk_classes c H) x y} ∈ c :=
+let ⟨b, hc, hy, hb⟩ := H y in eq_eqv_class_of_mem H hc hy ▸ hc
+
+/-- The union of a set of sets covering α is the entirety of α. -/
+lemma eqv_classes_union {c : set (set α)} (H : ∀ a, ∃ b ∈ c, a ∈ b) :
+  set.sUnion c = @set.univ α :=
+set.univ_subset_iff.1 $ λ x hx, let ⟨b, hm, hb⟩ := H x in set.mem_sUnion_of_mem hb hm
+
+/-- Distinct elements of a set of sets partitioning α are disjoint. -/
+lemma eqv_classes_disjoint {c : set (set α)}
+  (H : ∀ a, ∃ b ∈ c, a ∈ b ∧ ∀ b' ∈ c, a ∈ b' → b = b')
+  (b₁ b₂) (h₁ : b₁ ∈ c) (h₂ : b₂ ∈ c) (h : b₁ ≠ b₂) : disjoint b₁ b₂ :=
```

Comment:
This entire line could be written (up to minor ordering differences) `: pairwise_disjoint c`. Similarly for the next 3 lemmas.